File: VLQ.vb
Public Module VLQ
    Private Function concat(ByVal str As String()) As String
        Dim s As String = ""
        For Each st As String In str
            s &= st
        Next
        Return s
    End Function
    Public Function ToVLQ(ByVal Value As Integer) As Byte()
        Dim b = New Binary().NewInteger(Value)
        Dim l As New List(Of String)
        Dim bb = b.Split7
        For i = 0 To bb.Count - 1
            If i < bb.Count - 1 Then
                l.Add("1" & bb(i))
            Else
                l.Add("0" & bb(i))
            End If
        Next
        b = New Binary().NewString(concat(l.ToArray))
        Return b.ToBytes
    End Function
    Public Function FromVLQ(ByVal Value As Byte()) As Integer
        Dim l As New List(Of String)
        For Each i As Byte In Value
            l.Add(Convert.ToString(i, 2))
        Next
        Dim b = New Binary().NewString(concat(l.ToArray))
        l.Clear()
        For Each bb As String In b.Split8
            l.Add(bb.Substring(1, 7))
        Next
        Return Convert.ToInt32(concat(l.ToArray), 2)
    End Function
    Public Function FromsVLQ(ByVal Value As Byte()) As Integer
        Dim l As New List(Of String)
        For Each i As Byte In Value
            l.Add(Convert.ToString(i, 2))
        Next
        Dim b = New Binary().NewString(concat(l.ToArray))
        l.Clear()
        For Each bb As String In b.Split8
            l.Add(bb.Substring(1, 7))
        Next
        Return ((Convert.ToInt32(concat(l.ToArray), 2)) + 0) \ 2 '1) * 2
    End Function
    Public Function TosVLQ(ByVal Value As Integer) As Byte()
        Value *= 2
        'Value -= 1
        Dim b = New Binary().NewInteger(Value)
        Dim l As New List(Of String)
        Dim bb = b.Split7
        For i = 0 To bb.Count - 1
            If i < bb.Count - 1 Then
                l.Add("1" & bb(i))
            Else
                l.Add("0" & bb(i))
            End If
        Next
        b = New Binary().NewString(concat(l.ToArray))
        Return b.ToBytes
    End Function
End Module
Public Class Binary
    Public Property Bits As String
    Function NewLong(ByVal Value As Long) As Binary
        Bits = Convert.ToString(Value, 2)
        Return Me
    End Function
    Function NewInteger(ByVal Value As Integer) As Binary
        Bits = Convert.ToString(Value, 2)
        Return Me
    End Function
    Function NewUInteger(ByVal Value As UInteger) As Binary
        Bits = Convert.ToString(Value, 2).PadLeft(32, "0"c)
        Return Me
    End Function
    Function NewByte(ByVal Value As Byte) As Binary
        Bits = Convert.ToString(Value, 2).PadLeft(8, "0"c)
        Return Me
    End Function
    Function NewSByte(ByVal Value As SByte) As Binary
        Bits = Convert.ToString(Value, 2)
        Return Me
    End Function
    Function NewString(ByVal Value As String) As Binary
        Bits = Value
        Return Me
    End Function
    Public Function ToBytes() As Byte()
        Dim l As New List(Of Byte)
        For Each s As String In Split8()
            If s = "00000000" Then Continue For
            l.Add(Convert.ToByte(s, 2))
        Next
        Return l.ToArray
    End Function
    Public Function Split7() As String()
        Dim pad As Integer = 7 - (Bits.Count Mod 7)
        Dim l As New List(Of String)
        Dim str As String = Bits.PadLeft(Bits.Length + pad, "0"c)
        For i = 0 To str.Count - 1 Step 7
            l.Add(str.Substring(i, 7))
        Next
        Return l.ToArray
    End Function
    Public Function Split8() As String()
        Dim pad As Integer = 8 - (Bits.Count Mod 8)
        Dim l As New List(Of String)
        Dim str As String = Bits.PadLeft(Bits.Length + pad, "0"c)
        For i = 0 To str.Count - 1 Step 8
            l.Add(str.Substring(i, 8))
        Next
        Return l.ToArray
    End Function
    Public Function IsLastVLQByte() As Boolean
        If Bits(0) = "0"c Then
            Return True
        Else
            Return False
        End If
    End Function
End Class

File: VLQ.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
public static class VLQ
{
	public static byte[] ToVLQ(long Value)
	{
		dynamic b = new Binary().NewLong(Value);
        string s = "";
		string[] bb = b.Split7;
		for (int i = 0; i <= bb.Length - 1; i++) {
            if (i < bb.Length - 1)
            {
				s += "1" + bb[i];
			} else {
				s += "0" + bb[i];
			}
		}
		b = new Binary().NewString(s);
		return b.ToBytes;
	}
	public static byte[] TosVLQ(long Value)
	{
        long v = Value;
        bool neg = v < 0;
        v *= 2;
        if (neg) {
            v=- 1;
            v *= -1;
        }
        return ToVLQ(v);
	}
	public static ulong FromVLQ(byte[] Value)
	{
        string s = "";
        foreach (byte i in Value)
        {
            s += Convert.ToString(i, 2);
        }
        Binary b = new Binary().NewString(s);
        s = "";
        foreach (string bb in b.Split8())
        {
            s += bb.Substring(1, 7);
        }
        return Convert.ToUInt64(s, 2);
	}
	public static long FromsVLQ(byte[] Value)
	{
        ulong vlq = FromVLQ(Value);
        bool neg = vlq % 2 == 1;
        if (neg) vlq++;
        vlq /= 2;
        if (neg) return -(long)vlq;
        return (long)vlq;
	}
}
public class Binary
{
	public string Bits { get; set; }
	public Binary NewLong(long Value)
	{
		Bits = Convert.ToString(Value, 2);
		return this;
	}
	public Binary NewInteger(int Value)
	{
		Bits = Convert.ToString(Value, 2);
		return this;
	}
	public Binary NewUInteger(uint Value)
	{
		Bits = Convert.ToString(Value, 2).PadLeft(32, '0');
		return this;
	}
	public Binary NewByte(byte Value)
	{
		Bits = Convert.ToString(Value, 2).PadLeft(8, '0');
		return this;
	}
	public Binary NewSByte(sbyte Value)
	{
		Bits = Convert.ToString(Value, 2);
		return this;
	}
	public Binary NewString(string Value)
	{
		Bits = Value;
		return this;
	}
	public byte[] ToBytes()
	{
		List<byte> l = new List<byte>();
		foreach (string s in Split8()) {
			if (s == "00000000")
				continue;
			l.Add(Convert.ToByte(s, 2));
		}
		return l.ToArray();
	}
	public string[] Split7()
	{
        int pad = 7 - (Bits.Length % 7);
		List<string> l = new List<string>();
		string str = Bits.PadLeft(Bits.Length + pad, '0');
        for (int i = 0; i <= str.Length - 1; i += 7)
        {
			l.Add(str.Substring(i, 7));
		}
		return l.ToArray();
	}
	public string[] Split8()
	{
		int pad = 8 - (Bits.Length % 8);
		List<string> l = new List<string>();
		string str = Bits.PadLeft(Bits.Length + pad, '0');
        for (int i = 0; i <= str.Length - 1; i += 8)
        {
			l.Add(str.Substring(i, 8));
		}
		return l.ToArray();
	}
	public bool IsLastVLQByte()
	{
		if (Bits[0] == '0') {
			return true;
		} else {
			return false;
		}
	}
}
